lib/internal/streams/async_iterator.js

```diff
@@ -0,0 +1,132 @@
+  if (resolve !== null) {
+    const data = iter.stream.read();
+    // we defer if data is null
+    // we can be expecting either 'end' or
+    // 'error'
+    if (data !== null) {
+      resolve(new AsyncIteratorRecord(data, false));
+      iter[kLastResolve] = null;
+      iter[kLastReject] = null;
+    }
+  }
+}
+
+function onReadable(iter) {
+  // we wait for the next tick, because it might
+  // emit an error with process.nextTick
+  process.nextTick(readAndResolve, iter);
+}
+
+function onEnd(iter) {
+  const resolve = iter[kLastResolve];
+  if (resolve !== null) {
+    resolve(new AsyncIteratorRecord(null, true));
+    iter[kLastResolve] = null;
+    iter[kLastReject] = null;
+  }
+  iter[kEnded] = true;
+}
+
+function onError(iter, err) {
+  const reject = iter[kLastReject];
+  // reject if we are waiting for data in the Promise
+  // returned by next() and store the error
+  if (reject !== null) {
+    reject(err);
+    iter[kLastResolve] = null;
+    iter[kLastReject] = null;
+  }
+  iter.error = err;
+}
+
+class ReadableAsyncIterator {
+  constructor(stream) {
+    this.stream = stream;
+    this[kLastResolve] = null;
+    this[kLastReject] = null;
+    this[kError] = null;
+    this[kEnded] = false;
+    this[kLastPromise] = null;
+
+    stream.on('readable', onReadable.bind(null, this));
+    stream.on('end', onEnd.bind(null, this));
+    stream.on('error', onError.bind(null, this));
+
+    // the function passed to new Promise
+    // is cached so we avoid allocating a new
+    // closure at every run
+    this[kHandlePromise] = (resolve, reject, promise) => {
+      // if we have multiple next() calls
+      // we will wait for the previous Promise to finish
+      const lastPromise = this[kLastPromise];
+      if (promise === lastPromise) {
+        this[kLastPromise] = null;
+      } else if (promise === undefined && lastPromise !== null) {
+        const boundFn =
+          this[kHandlePromise].bind(this, resolve, reject, lastPromise);
+        lastPromise.then(boundFn, boundFn);
+        return;
+      }
+
+      const data = this.stream.read();
+      if (data) {
+        resolve(new AsyncIteratorRecord(data, false));
+      } else {
+        this[kLastResolve] = resolve;
+        this[kLastReject] = reject;
+      }
+    };
+  }
+
+  next() {
+    // if we have detected an error in the meanwhile
+    // reject straight away
+    const error = this[kError];
+    if (error !== null) {
+      return Promise.reject(error);
+    }
+
+    if (this[kEnded]) {
+      return Promise.resolve(new AsyncIteratorRecord(null, true));
+    }
+
+    const promise = new Promise(this[kHandlePromise]);
+    this[kLastPromise] = promise;
+    return promise;
+  }
+
+  return() {
+    // destroy(err, cb) is a private API
+    // we can guarantee we have that here, because we control the
+    // Readable class this is attached to
+    return new Promise((resolve, reject) => {
+      this.stream.destroy(null, (err) => {
```

Comment:
Just thinking about this a bit more, I think we might need to unset `lastPromise` if it's not `null` here. It won't do anything weird because there's no way it can get resolved anymore but I don't think it will get GC'd either since it's unresolved.